### ISSUE TITLE
Implement Waddah attar explosion

### ISSUE THREAD

**Issue by jbax** [uniVocity/univocity-trader#12]
title

**Comment by ChrisTrakas** [comment]
Hello , are you still interested in the implementation of this signal?

**Comment by jbax** [comment]
Yes

**Comment by ChrisTrakas** [comment]
So , I'm trying to implement WAE , but this indicator needs to access previous values. To be specific , a WAE object will cointain one MACD and one BollingerBand object. In the calculate method of WAE though , I need to get the MACD and BollingerBand values of the previous 3 candles. Any ideas on how I could work that out?

**Comment by jbax** [comment]
Thanks for the help! Use a CircularList of size 3 to store the values of
each indicator. It will discard older values.

**Comment by ChrisTrakas** [comment]
Sure, but how could I use the old and current values simultaneously?

**Comment by jbax** [comment]
I'm not sure what you mean exactly, if you aren't sure about how to use it
take a look at some indicator that uses circularlist:

https://github.com/uniVocity/univocity-trader/blob/master/univocity-trader-core/src/main/java/com/univocity/trader/indicators/AwesomeOscillator.java#L46

You can probably go without even using a circularlist, as you only need 3
recent values, just store them in a few attributes, like done here:

https://github.com/uniVocity/univocity-trader/blob/master/univocity-trader-core/src/main/java/com/univocity/trader/indicators/AwesomeOscillator.java#L46

Hope this helps.

**Comment by jbax** [comment]
Ops I meant to link this for the example that uses circularlist:
https://github.com/uniVocity/univocity-trader/blob/master/univocity-trader-core/src/main/java/com/univocity/trader/indicators/InstantaneousTrendline.java#L14

**Comment by ChrisTrakas** [comment]
Okay I will check them out. What I meant was , lets say that you want to get the output of the WAE in the current moment. To calculate this , you need to use the MACD values from the past 3 candles. If for example we are in the 1 hour timeframe , to get the WAE value for 15.00 we also need to use the MACD values from 14.00 , 13.00 , 12.00 . I haven't figured out how to do that yet.

**Comment by jbax** [comment]
The examples should help you. First keep in mind all calculations are
performed on the fly and there is no internal list of candles you can
navigate. Unless you explicitly add stuff to circularlists.

Then you'll probably want to extend from SingleValueIndicator and override
the process method (I'm on mobile so sorry for the bad formatting) :

protected boolean process(Candle candle, double value, boolean updating) {
if (updating) {
 //if your indicator reacts to every tick until a full 1h candle is
produced, the most recent value in the list will be updated
Macd.accumulate(candle)
Macdcircularlist.update(macd. Getvalue() );
} else {
// otherwise you add the vale discarding the oldest
Macdcircularlist.add(macd. Getvalue() );}

...
Calculate the indicator. E. G

MacdcircularList.Getrecentvalue(3) will get the value oldest macd value
accumulated if your circularlist is of size 3.

Does that help?

**Comment by ChrisTrakas** [comment]
Thanks ,  I got it , the circular list will solve the problem.

**Comment by jbax** [comment]
Thanks for your PR, I made a few adjustments to make the calculation work correctly. I also added the necessary code to make it display on a chart on the test code I'm putting up together in class `ChartWindow` (right click on a chart and select "indicators" to see it). I still need to get it to show up the correct histogram colors. Will hopefully complete that tomorrow. 

I'll leave this issue open until I finalize this and also ensure the code handles tick updates to make the last signal of the indicator be updated every time a new tick comes for the last bar.
